### PR TITLE
[feat] Add AMPnet implementation in experimental dir

### PR DESCRIPTION
**Pull request by sidgoyal78**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
This PR adds the implementation of AMPnet (https://arxiv.org/abs/1705.09786). We replicate the MNIST results from the paper and add a benchmark on identity prediction LM task using transformers.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/pipe_process/test_pipe.py [hunk 1]
```diff
@@ -25,6 +25,8 @@
 import pytest
 import torch
 from torch import nn
+from torch.optim.optimizer import Optimizer
```
Comment (min-xu-ai) [inline tests/nn/pipe_process/test_pipe.py:28]: does it make sense to have the new tests in a new file?

File: experimental/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+__version__ = "0.1.4"
```
Comment (min-xu-ai) [inline experimental/__init__.py:6]: is this needed?

Can you add some high level doc on:

1. what's this variant of pipeline
2. how to use it

Thanks for moving to the experimental dir. Really appreciate it. I think adding a bit doc and I think it is ready to go. @msbaines @blefaudeux, @anj-s what do you guys think?

File: tests/nn/pipe_process/test_pipe.py (hunk 1 above)
Comment (sidgoyal78) [reply tests/nn/pipe_process/test_pipe.py:28]: Yeah, good point. I will move new tests under experimental.

File: experimental/__init__.py (hunk 2 above)
Comment (sidgoyal78) [reply experimental/__init__.py:6]: Yes, I will add a markdown file with results. 
We're still in the process of getting some more results. But let me add the current set.

Comment (sidgoyal78) [reply experimental/__init__.py:6]: > is this needed?
I don't think so. Thanks for pointing it out.

File: tests/nn/pipe_process/test_pipe.py (hunk 1 above)
Comment (sidgoyal78) [reply tests/nn/pipe_process/test_pipe.py:28]: @min-xu-ai : Do you suggest adding these 2 tests in a new file in the same location? Something like tests/nn/pipe_process/test_ampnet_pipe.py

Or do you think we should create a new directory tests under experimental? Something like experimental/tests/nn/pipe_process/test_ampnet_pipe.py

Comment (sidgoyal78) [reply tests/nn/pipe_process/test_pipe.py]: As of the recent-most commit, I followed the second option. And we can run those tests using `pytest experimental/tests`

But happy to do it the other way.

Comment (min-xu-ai) [reply tests/nn/pipe_process/test_pipe.py]: sounds good to me. thanks a lot!

File: experimental/__init__.py (hunk 2 above)
Comment (sidgoyal78) [reply experimental/__init__.py]: @min-xu-ai I added a docstring in the class AMPnetPipe. I also added instructions on how to run the benchmark script, experimental_ampnet.py

**Comment by sidgoyal78** [comment]
@msbaines and @min-xu-ai : Could you please take another look? I refactored the code further, added documentation.

**Review by msbaines — APPROVED** [review]

**Pull request merged**